clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1064,6 +1075,77 @@ static target getAccessTarget(QualType FieldTy,
       AccTy->getTemplateArgs()[3].getAsIntegral().getExtValue());
 }
 
+static bool isFreeFunction(SemaSYCL &SemaSYCLRef, const FunctionDecl *FD) {
+  for (auto *IRAttr : FD->specific_attrs<SYCLAddIRAttributesFunctionAttr>()) {
+    SmallVector<std::pair<std::string, std::string>, 4> NameValuePairs =
+        IRAttr->getAttributeNameValuePairs(SemaSYCLRef.getASTContext());
+    for (const auto &NameValuePair : NameValuePairs) {
+      if (NameValuePair.first == "sycl-nd-range-kernel" ||
+          NameValuePair.first == "sycl-single-task-kernel")
+        return true;
+    }
+  }
+  return false;
+}
+
+static std::string constructFFKernelName(const FunctionDecl *FD) {
+  IdentifierInfo *Id = FD->getIdentifier();
+  std::string NewIdent = (Twine("__sycl_kernel_") + Id->getName()).str();
+  return NewIdent;
+}
+
+// Creates a name for the free function kernel function.
+// Non-templated functions get a simple __sycl_kernel_ prefix.
+// For templated functions we add __sycl_kernel_ to the original function name
+// and then use the mangled name as the kernel name. The renaming allows a
+// normal device function to coexist with the kernel function.
+static std::pair<std::string, std::string> constructFreeFunctionKernelName(
+    SemaSYCL &SemaSYCLRef, const FunctionDecl *FreeFunc, MangleContext &MC) {
+  SmallString<256> Result;
+  llvm::raw_svector_ostream Out(Result);
+  std::string MangledName;
+  std::string StableName;
+
+  if (FreeFunc->getTemplateSpecializationArgs()) {
+    ASTContext &Ctx = SemaSYCLRef.getASTContext();
+    FunctionProtoType::ExtProtoInfo Info(CC_OpenCLKernel);
+    QualType FuncType = Ctx.getFunctionType(Ctx.VoidTy, {}, Info);
+    std::string FFName =
+        (Twine("__sycl_kernel_") + FreeFunc->getIdentifier()->getName()).str();
+    const IdentifierInfo *FFIdent = &Ctx.Idents.get(FFName);
+    FunctionDecl *NewFD = FunctionDecl::Create(
```

Comment:
This seems like overkill to get the mangled name. Maybe we need to be writing our own mangling function? This honestly feels like a hack to me.

Reply:
Perhaps overkill, but a reliable way to get a name for the free function kernel. Can you suggest a simpler alternative?

Reply:
I don't have experience with this and so I don't know off the top of my head.  @tahonermann @AaronBallman can you weigh in here please?

Reply:
I should mention that this function is curently used only for generating a name but in a future PR the created FunctionDecl will be used in some declarations in the integration header for implementing the get_kernel_id API and free function traits.

Reply:
Why would you need this `FunctionDecl` for that? I would assume you can use the FunctionDecl actually corresponding to the kernel in that case.

Also, I don't know anything about free function design limitations but we really should not be designing things around the integration header anymore considering we cannot upstream it.

Reply:
Note that from one user-written function we generate two functions. The original function and its FunctionDecl remain the same. The new one reflects a renaming, and is for the kernel function version.
The first version of free function implementation will use the integration header. Later, if really necessary, we can use an alternative design.

Reply:
This FunctionDecl is created solely for the purposes of generating a unique name. That's why we use the original function as a basis, and only change its identifier to give it a __sycl_kernel prefix. If the original function's instantiated overloaded/instantiated versions were unique, then the naming scheme we have used will generate unique names for the kernel functions as well.

Reply:
I do think it would be best if we can avoid creating multiple `FunctionDecl`s for use with the generated kernel. Ideally, the mangled name would match the parameters for the generated kernel, but that isn't strictly necessary as long as names remain unique. Deviations could cause some confusion though.

Reply:
I think the issue is you need a `FunctionDecl` for creating the mangled name and we don't have a `FunctionDecl` with name '__sycl_kernel_ff' (the name is just ff). So there is this weird hack of creating a temporary FD with name '__sycl_kernel_ff' just to get its mangled name, which we then use to create the actual kernel `FunctionDecl` . It seems hacky to me but I also don't know how else to do it other than rewriting mangling functions to not require a FD. I have not explored that though. 

>>I do think it would be best if we can avoid creating multiple FunctionDecls for use with the generated kernel. Ideally, the mangled name would match the parameters for the generated kernel, but that isn't strictly necessary as long as names remain unique. Deviations could cause some confusion though.

Yea whenever lowering is not straightforward (as in the case of special types and pointers) there are going to be deviations.

Reply:
I agree with Tom, we should not create an AST node just to get a unique name from it. You should be able to mangle the original `FunctionDecl` to get a mangled name, then do a (custom) string replace on the original function name based on which mangling scheme is being used. It's a bit annoying, but it's better than increasing memory pressure by making AST nodes that won't ever be needed for the AST.

Reply:
Can the mangled name be produced as part of constructing `KernelDecl` in `SyclKernelDeclCreator`? This is what I meant by avoiding creating multiple `FunctionDecl`s.

Reply:
Aaron's string replacement suggestion should work ok (assuming free functions are limited to functions that have a name; e.g., no overloaded operator or user defined literal functions. That string replacement approach would also have to substitute for string lengths embedded in the name. e.g., `void foo()` in the IA64 ABI would be mangled as `_Z3foov` where the `3` is the length of the identifier `foo`.

My preference is to construct a proper mangled name using the `FunctionDecl` constructed for the kernel function if that is reasonably possible (which I think it should be).

Reply:
> My preference is to construct a proper mangled name using the `FunctionDecl` constructed for the kernel function if that is reasonably possible (which I think it should be).

If we can go that route, that's my preference as well.

Reply:
I was trying to avoid a string replacement approach, but that seems better than creating a special FunctionDecl.

There are issues with using the FunctionDecl of the kernel function as a source of the name. Firstly, to create the FunctionDecl we need a name, and the proposal is to create a name from the FunctionDecl. That's a catch22. Secondly, the original free function is guaranteed to have a unique name and prepending __sycl_kernel will also yield a unique name. However, a kernel function may have expanded parameters, e.g. for an accessor we would get a pointer and two ranges as replacements for the accessor. It is possible that a second free function (overloaded) with the same name would have a pointer and two ranges in place of the accessor parameter. Now, names constructed from the kernel function would conflict.

Reply:
I don't see the catch-22. The kernel function and the free function are distinct `FunctionDecl`s and the latter is needed to generate the former. However, ...

Good catch on the possibility of a name conflict due to decomposition. I agree that could happen and that does indeed mean that we can't just drive creation of the mangled name based on the constructed kernel decl. As much as the string replacement approach feels like a hack, it should work well. Maybe that is the best option, at least for the near term.

Reply:
Couldn't you create a `FunctionDecl` for the kernel function *without* decomposing the arguments?  You would prefix the function name with "\_\_sycl\_kernel\_", and then use this `FunctionDecl` only for generating the mangled name.  Obviously, you would still do argument decomposition when generating the body of the kernel function.

Does the "string replacement algorithm" mean that you start with the mangled name and then try to add a prefix to the "root" part of the function?  This is not really my area of expertise, but it seems like that would either be fragile or require intimate knowledge of the mangling algorithm.  Remember that the kernel function might be nested in a bunch of namespaces.  In the future, we want to allow kernel functions that are static member functions of a class.  It seems like it would be hard to find the part of the string that represents the "root" of the function name without understanding the details of the mangling algorithm.

Reply:
"Couldn't you create a FunctionDecl for the kernel function without decomposing the arguments?" Yes, that is one approach being pursued now. It needs one additional change, to account for template specialization parameters, which isn't working yet.

The other approach, hunting for the function name in a mangled string will be tricky. That's why I didn't go down that road initially. But in limited scenarios that could be implemented.

Reply:
In this PR I would like to use the string manipulation approach in function constructFreeFunctionKernelName. It is known to work for free functions at file scope, which is all we support right now. The kernel names are the same for Linux and Windows. This is in line with normal SYCL kernels that also use the same names on Linux and Windows.

A future PR might use a different scheme for generating the free function kernel name.